File: AdventOfCode22/Puzzle1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {

	// Open the input file
	readFile, err := os.Open("./input.txt")
	check(err)
	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)

	var currentLine string
	var calorieLine, calorieValue int
	var calorieList []int

	// To find the top N elves we need an N+1 array
	highestCalories := []int{0, 0, 0, 0}

	for fileScanner.Scan() {
		// If the next line is only a newline character,
		// assume that everything before is carried by the same elf.
		currentLine = fileScanner.Text()

		if currentLine == "" {
			// Sum the list of items carried by an elf
			calorieValue = sumArray(calorieList)

			// Insert the value into the array in the "dummy" slot
			highestCalories[0] = calorieValue

			// Determine the 3 highest calorie elves from the 4 values
			highestCalories = compareCalorieCount(highestCalories)

			// Reset the running total
			calorieList = nil
		} else {
			calorieLine, err = strconv.Atoi(currentLine)
			check(err)
			calorieList = append(calorieList, calorieLine)
		}
	}
	// Close the input file
	readFile.Close()

	// Puzzle 1: Part 1: https://adventofcode.com/2022/day/1
	fmt.Printf("Most caloric elf is carrying [%d] calories.\n", highestCalories[3])

	// Puzzle 1: Part 2: https://adventofcode.com/2022/day/1#part2
	// Sum the array of highest calories (the 4th value is always 0 so it's only summing 3)
	calorieValue = sumArray(highestCalories)
	fmt.Printf("Calories carried by top 3 elves = [%d] calories.\n", calorieValue)
}

func sumArray(values []int) int {
	// Sum the values of an array
	calorieCount := 0
	for _, v := range values {
		calorieCount += v
	}
	return calorieCount
}

func compareCalorieCount(values []int) []int {
	// Sort the values from low to high and remove (zero) the lowest one
	sort.Ints(values)
	values[0] = 0
	return values
}

File: AdventOfCode22/Puzzle1/rust/src/main.rs
use std::fs;

fn main() {
    // Open the input file
    let file_path = r"D:\Git\bits-and-pieces\AdventOfCode22\Puzzle1\input.txt";
    let contents = fs::read_to_string(file_path).expect("Should have been able to read the file");

    let mut calorie_value: u32 = 0;
    let mut calorie_list: Vec<u32> = Vec::new();

    for line in contents.lines() {
        match line.parse::<u32>() {
            Ok(n) => {
                // Add the value of the line to the current calorie value
                calorie_value += n;
            }
            _ => {
                // If the line was not able to be parsed to an int, then it must be an empty line
                calorie_list.push(calorie_value);
                // Reset the calorie counter
                calorie_value = 0;
            }
        }
    }

    // Sort Ascending
    calorie_list.sort();
    // Sort Descending (easier to slice)
    calorie_list.reverse();

    // Part One
    println!("Part One: Calories of highest elf: {:?}", calorie_list[0]);

    // Part Two
    let sum: u32 = calorie_list[0..=2].iter().sum();
    println!("Part Two: Calories of top 3 elves: {:?}", sum);
}
